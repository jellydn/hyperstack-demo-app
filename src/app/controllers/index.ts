// naming and imports: name and import as you see fit
import Auth from './auth'
import { NotesController } from './notes'

const controllers = []
controllers.push(Auth)
controllers.push(NotesController)

export const AppController = {
  controllers,
}
