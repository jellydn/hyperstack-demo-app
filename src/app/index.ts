import type { Context } from 'hyperstack'
import { context } from 'hyperstack'
import type * as Models from './models'
import type * as Mailers from './mailers'

export const appContext = context as Context<typeof Models, typeof Mailers>
