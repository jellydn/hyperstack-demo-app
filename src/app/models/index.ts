export { Note } from './note'
export { User } from './user'
