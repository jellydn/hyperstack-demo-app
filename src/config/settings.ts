// general purpose settings, with no infrastructure loading, no assumptions. keep it simple.
import path from 'path'
export const root = path.join(__dirname, '..')
