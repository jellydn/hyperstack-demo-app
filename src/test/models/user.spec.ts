import { test } from '@hyperstackjs/testing'
import L from 'lodash'
import { getProps } from '@hyperstackjs/initializer-jwt'
import { root } from '../../config/settings'
import { appContext } from '../../app'
import { baseFixture } from '../__fixtures__'

const { models } = test(root)

const serializeModel = (m: any) => {
  return L.omit(m.toJSON(), ['pid'])
}

describe('models', () => {
  describe('user', () => {
    models('should create and login', async (_app) => {
      const { verifyJWT } = getProps()
      const { User } = appContext.models()
      const u = await User.createWithPassword({
        username: 'evh@example.com',
        name: 'Eddie Van Halen',
        password: 'jump!$',
      })
      expect(serializeModel(u)).toMatchSnapshot()
      expect(await u.verifyPassword('jump!$')).toEqual(true)
      expect(u.pid).toMatch(/.+-.+-.+-.+-.+/)
      expect(u.password).toMatch(/^pbkdf2_sha256/)
      expect(u.createAuthenticationToken()).toMatch(/^eyJh/)
      expect((await verifyJWT(u.createAuthenticationToken())).sub).toEqual(
        'evh@example.com'
      )
    })
    models('should reset password', async () => {
      await baseFixture()
      const { User } = appContext.models()
      const alex = await User.findOne({
        where: { username: 'alex@example.com' },
      })
      if (!alex) {
        throw new Error('user is null')
      }
      expect(alex.resetPassword('')).rejects.toThrow(/too short/)
      await alex.resetPassword('some-other-password')
      await alex.reload()
      expect(await alex.verifyPassword('alex-might-password')).toBeFalse()
      expect(await alex.verifyPassword('some-other-password')).toBeTrue()
    })
    models('should set forgot token', async () => {
      await baseFixture()
      const { User } = appContext.models()
      const alex = await User.findOne({
        where: { username: 'alex@example.com' },
      })
      if (!alex) {
        throw new Error('user is null')
      }
      expect(alex.resetToken).toBeNull()
      expect(alex.resetSentAt).toBeNull()
      await alex.forgotPassword()
      expect(alex.resetToken).toBeTruthy()
      expect(alex.resetSentAt).toBeTruthy()
    })
    models('should verify account', async () => {
      await baseFixture()
      const { User } = appContext.models()
      const alex = await User.findOne({
        where: { username: 'alex@example.com' },
      })
      if (!alex) {
        throw new Error('user is null')
      }
      expect(alex.emailVerificationToken).toBeTruthy()
      expect(alex.emailVerificationSentAt).toBeTruthy()
      expect(alex.emailVerifiedAt).toBeNull()
      await alex.verified()
      expect(alex.emailVerificationToken).toBeNull()
      expect(alex.emailVerificationSentAt).toBeNull()
      expect(alex.emailVerifiedAt).toBeTruthy()
    })
  })
})
