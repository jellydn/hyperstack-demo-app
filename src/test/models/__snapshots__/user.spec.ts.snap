// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models user should create and login 1`] = `
{
  "name": "Eddie Van Halen",
  "username": "evh@example.com",
}
`;
